feat: explicitly exported component types

EcksDy opened the pull request:
I've started poking around the codebase to see what needs to be done to enable this, and this seems fairly simple on the surface.

Parse the properties of the schema interface, map the names to the type alias and reference, export while making sure no named exports are colliding, and lastly add at the end of the `schema.ts` file.

So each property in an interface such as this one:
```ts
export interface components {
    schemas: {
        SomeExampleDto: {
            createdAt?: string;
        };
        // ...
    };
    // ...
}
```

Becomes the following export at the end of file:
```ts
export type SomeExampleDto = components['schemas']['SomeExampleDto'];
```

I've taken a naive approach at the AST. Need help with testing that properly, not sure how to go about AST fixtures.

radist2s commented:
@EcksDy, I would not waste time testing AST. It's much easier to add integration tests with `--explicit-component-exports` option and match the snapshot with the expected output. I've added a new test case that checks the option. Your feature looks cool.

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+export function createExportsForComponentSchemas(ast: ts.Node[]) {
+  const componentsInterface = ast.find(
+    (n) => isInterfaceDeclaration(n) && n.name.escapedText === 'components'
+  ) as ts.InterfaceDeclaration | undefined;
```
Using following code you can avoid using `as`:
```ts
const componentsInterface = ast
    .filter(isInterfaceDeclaration)
    .find((n) => n.name.escapedText === 'components');

// componentsInterface will be iferred as `InterfaceDeclaration`
```

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+      isPropertySignature(n) &&
+      isIdentifier(n.name) &&
+      n.name.escapedText === 'schemas'
+  ) as ts.PropertySignature | undefined;
```
The same approach as with `isInterfaceDeclaration(...)` could be used to avoid `as` casting.

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+
+function getExportedNames(ast: ts.Node[]) {
+  return ast.reduce<string[]>((acc, curr) => {
+    const hasExportModifier = (curr as any).modifiers?.some(isExportKeyword);
```
I would refactor this to more type-safe code:
```ts
const hasExportModifier =
      'modifiers' in curr &&
      Array.isArray(curr.modifiers) &&
      curr.modifiers?.some(isExportKeyword);
```

EcksDy commented:
@radist2s thank you, my pleasure.

I've removed the overkill test in favor of the snapshot testing you've introduced.

On another topic, there might be a potential bug here:
![image](https://github.com/user-attachments/assets/08fde35d-bb41-47ef-9b43-9763537b8b3e)
but the option has a different shape:
![image](https://github.com/user-attachments/assets/feb2e0fc-1518-4e9c-8ce3-8986ae14e2f5)

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+  return exportStatements;
+}
+
+function getExportedNames(ast: ts.Node[]) {
```
I would refactor the function like that:
```ts
function getExportedNames(ast: ts.Node[]) {
  return ast.reduce<string[]>((acc, node) => {
    if (
      ts.isInterfaceDeclaration(node) &&
      ts.isIdentifier(node.name) &&
      node.modifiers?.some(
        (modifier) => modifier.kind === ts.SyntaxKind.ExportKeyword
      )
    )
      acc.push(node.name.text.toString());

    return acc;
  }, []);
}
```

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+        (m) =>
+          isPropertySignature(m) && isIdentifier(m.name) && m.name.escapedText
+      )
+      .filter((name) => !!name && !exportedNames.includes(name)) as string[]);
```
I would avoid casting here too. Btw, `name` is not a 'string':

```ts
  const schemaNames =
  componentSchemas.type &&
  ts.isTypeLiteralNode(componentSchemas.type) &&
  componentSchemas.type.members.reduce<string[]>((acc, element) => {
    if (ts.isPropertySignature(element) && ts.isIdentifier(element.name)) {
      const componentName = element.name.text.toString();
      if (!exportedNames.includes(componentName)) acc.push(componentName);
    }
    return acc;
  }, []);
```

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+  ) as ts.PropertySignature | undefined;
+  if (!componentSchemas) return [];
+
+  const exportedNames = getExportedNames(ast);
```
It seems that `getExportedNames(...)` implementation should be tested. The test must check that the function returns anything, because if function returns nothing, the current test will not fail.
I recommend adding a test and compare output with `toMatchInlineSnapshot(...)` for simplicity.

EcksDy replied:
Hmmm, not sure what you mean, it behaves correctly if `exportedNames` are empty.
`exportedNames` are an exclusion list to make sure that the generated exports do not overlap with any existing exports. 

I've added it because in my usecase the schemas included an enum and I had an `openapi-typescript` enum flag that added explicit exports.

radist2s replied:
@EcksDy, pushed tests for `getExportedNames(...)`. There are no issues with the function implementation. But in my case, if I replace its result with an empty array, all tests still pass. If I accidentally change the implementation of the `getExportedNames(...)` tests will not fail.
The idea is to cover cases (with unit tests) that could not be tested using integration tests.

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+    if (!hasIdentifier) return acc;
+
+    const identifier = (curr as any).name as ts.Identifier;
+    acc.push(identifier.escapedText as string);
```
In TSDoc of the property I see that recommended using not escaped text:
> Prefer to use `id.unescapedText`. (Note: This is available only in services, not internally to the TypeScript compiler.)
> Text of identifier, but if the identifier begins with two underscores, this will begin with three.

As I see, there is `.text` property. Have you checked it?

radist2s commented on packages/openapi-typescript-plugin/src/lib/createExportsForComponentSchemas.ts:
```diff
@@ -0,0 +1,86 @@
+  return node?.kind === ts.SyntaxKind.PropertySignature;
+}
+
+function isIdentifier(node?: ts.Node): node is ts.Identifier {
```
In the most, TS has its own helpers such `ts.isIdentifier(...)` or `ts.isTypeLiteralNode(...)`. I would recommend using built-in helpers.

radist2s commented:
@EcksDy, according to `--no-blob-from-binary` works inversely: while it is not provided, `blobFromBinary` is `true`. So, there is no bug.

radist2s commented on packages/openapi-typescript-plugin/src/generateSchemaTypes.ts:
```diff
@@ -28,28 +29,32 @@ export async function generateSchemaTypes(
+    silent,
+  });
+
+  const componentSchemasExportNodes = createExportsForComponentSchemas(ast);
```
We could optimize it like this:
```ts
if (args.explicitComponentExports)
  return astToString(ast.concat(createExportsForComponentSchemas(ast)));

return astToString(ast);
```

EcksDy commented:
I liked the filter comments. I avoided filters initially to skip another iteration, but it's negligible. I've also removed all assertion.

`text` is string so I removed the casting.

radist2s reviewed (approved):
This looks like a great contribution! Thanks more for your enthusiasm. I will suggest this option to my team to use on Monday.

radist2s commented:
@EcksDy, please message me if PR is ready and I can merge.

EcksDy commented:
@radist2s sure, all done from my side I think.
Lets see where this goes, I might want to add something else in the future 😄
